File: src/lib/avatars.ts
import apple from "../assets/avatars/Apple.jpg";
import artichoke from "../assets/avatars/Artichoke.jpg";
import bread from "../assets/avatars/Bread.jpg";
import cheese from "../assets/avatars/Cheese.jpg";
import eggplant from "../assets/avatars/Eggplant.jpg";
import fig from "../assets/avatars/Fig.jpg";
import garlic from "../assets/avatars/Garlic.jpg";
import grape from "../assets/avatars/Grape.jpg";
import hazelnut from "../assets/avatars/Hazelnut.jpg";
import honey from "../assets/avatars/Honey.jpg";
import lemon from "../assets/avatars/Lemon.jpg";
import mushroom from "../assets/avatars/Mushroom.jpg";
import onion from "../assets/avatars/Onion.jpg";
import orange from "../assets/avatars/Orange.jpg";
import peach from "../assets/avatars/Peach.jpg";
import pear from "../assets/avatars/Pear.jpg";
import pepper from "../assets/avatars/Pepper.jpg";
import pomegranate from "../assets/avatars/Pomegranate.jpg";
import potato from "../assets/avatars/Potato.jpg";
import pumpkin from "../assets/avatars/Pumpkin.jpg";
import quince from "../assets/avatars/Quince.jpg";
import tomato from "../assets/avatars/Tomato.jpg";
import walnut from "../assets/avatars/Walnut.jpg";

export type Avatar = {
  id: string;
  name: string;
  src: string;
};

export const AVATARS: Avatar[] = [
  { id: "apple", name: "Apple", src: apple },
  { id: "artichoke", name: "Artichoke", src: artichoke },
  { id: "bread", name: "Bread", src: bread },
  { id: "cheese", name: "Cheese", src: cheese },
  { id: "eggplant", name: "Eggplant", src: eggplant },
  { id: "fig", name: "Fig", src: fig },
  { id: "garlic", name: "Garlic", src: garlic },
  { id: "grape", name: "Grape", src: grape },
  { id: "hazelnut", name: "Hazelnut", src: hazelnut },
  { id: "honey", name: "Honey", src: honey },
  { id: "lemon", name: "Lemon", src: lemon },
  { id: "mushroom", name: "Mushroom", src: mushroom },
  { id: "onion", name: "Onion", src: onion },
  { id: "orange", name: "Orange", src: orange },
  { id: "peach", name: "Peach", src: peach },
  { id: "pear", name: "Pear", src: pear },
  { id: "pepper", name: "Pepper", src: pepper },
  { id: "pomegranate", name: "Pomegranate", src: pomegranate },
  { id: "potato", name: "Potato", src: potato },
  { id: "pumpkin", name: "Pumpkin", src: pumpkin },
  { id: "quince", name: "Quince", src: quince },
  { id: "tomato", name: "Tomato", src: tomato },
  { id: "walnut", name: "Walnut", src: walnut },
];

const byId = new Map(AVATARS.map((avatar) => [avatar.id, avatar]));

export function avatarById(id: string | undefined): Avatar | undefined {
  if (!id) return undefined;
  return byId.get(id);
}

File: src-tauri/src/chat/avatars.rs
//! Conversation faces. Ids match `src/lib/avatars.ts`.

use std::collections::HashSet;

pub struct Avatar {
    pub id: &'static str,
    pub name: &'static str,
}

/// Spoken name is the food, not the filename.
pub const AVATARS: &[Avatar] = &[
    Avatar {
        id: "apple",
        name: "Apple",
    },
    Avatar {
        id: "artichoke",
        name: "Artichoke",
    },
    Avatar {
        id: "bread",
        name: "Bread",
    },
    Avatar {
        id: "cheese",
        name: "Cheese",
    },
    Avatar {
        id: "eggplant",
        name: "Eggplant",
    },
    Avatar {
        id: "fig",
        name: "Fig",
    },
    Avatar {
        id: "garlic",
        name: "Garlic",
    },
    Avatar {
        id: "grape",
        name: "Grape",
    },
    Avatar {
        id: "hazelnut",
        name: "Hazelnut",
    },
    Avatar {
        id: "honey",
        name: "Honey",
    },
    Avatar {
        id: "lemon",
        name: "Lemon",
    },
    Avatar {
        id: "mushroom",
        name: "Mushroom",
    },
    Avatar {
        id: "onion",
        name: "Onion",
    },
    Avatar {
        id: "orange",
        name: "Orange",
    },
    Avatar {
        id: "peach",
        name: "Peach",
    },
    Avatar {
        id: "pear",
        name: "Pear",
    },
    Avatar {
        id: "pepper",
        name: "Pepper",
    },
    Avatar {
        id: "pomegranate",
        name: "Pomegranate",
    },
    Avatar {
        id: "potato",
        name: "Potato",
    },
    Avatar {
        id: "pumpkin",
        name: "Pumpkin",
    },
    Avatar {
        id: "quince",
        name: "Quince",
    },
    Avatar {
        id: "tomato",
        name: "Tomato",
    },
    Avatar {
        id: "walnut",
        name: "Walnut",
    },
];

pub fn name_for(id: &str) -> Option<&'static str> {
    AVATARS
        .iter()
        .find(|avatar| avatar.id == id)
        .map(|a| a.name)
}

pub fn pick_id(thread_id: &str, used: &HashSet<String>) -> &'static str {
    let unused: Vec<&Avatar> = AVATARS
        .iter()
        .filter(|avatar| !used.contains(avatar.id))
        .collect();
    if unused.is_empty() {
        return AVATARS[hash_thread_id(thread_id) % AVATARS.len()].id;
    }
    unused[hash_thread_id(thread_id) % unused.len()].id
}

fn hash_thread_id(thread_id: &str) -> usize {
    let mut hash: u32 = 0;
    for byte in thread_id.bytes() {
        hash = hash.wrapping_mul(31).wrapping_add(u32::from(byte));
    }
    hash as usize
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn names_match_catalog_ids() {
        assert_eq!(name_for("tomato"), Some("Tomato"));
        assert_eq!(name_for("pomegranate"), Some("Pomegranate"));
        assert_eq!(name_for("nope"), None);
    }

    #[test]
    fn pick_skips_faces_already_in_use() {
        let used: HashSet<String> = AVATARS
            .iter()
            .skip(1)
            .map(|avatar| avatar.id.to_string())
            .collect();
        assert_eq!(pick_id("t_anything", &used), AVATARS[0].id);
    }

    #[test]
    fn pick_is_stable_when_every_face_is_taken() {
        let used: HashSet<String> = AVATARS.iter().map(|avatar| avatar.id.to_string()).collect();
        assert_eq!(pick_id("same-thread", &used), pick_id("same-thread", &used));
        assert!(name_for(pick_id("same-thread", &used)).is_some());
    }
}
